src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -819,26 +826,100 @@ private BoundExpression BindTupleExpression(TupleExpressionSyntax node, Diagnost
+                tupleTypeOpt = TupleTypeSymbol.Create(node.Location, elements, locations, elementNamesArray,
+                    this.Compilation, syntax: node, diagnostics: diagnostics, shouldCheckConstraints: true);
             }
             else
             {
                 TupleTypeSymbol.VerifyTupleTypePresent(elements.Length, node, this.Compilation, diagnostics);
             }
 
-            return new BoundTupleLiteral(node, elementNamesArray, boundArguments.ToImmutableAndFree(), tupleTypeOpt, hasErrors);
+            return new BoundTupleLiteral(node, elementNamesArray, inferredArray, boundArguments.ToImmutableAndFree(), tupleTypeOpt, hasErrors);
+        }
+
+        private static (ImmutableArray<string> names, ImmutableArray<bool> inferred) MergeTupleElementNamesAndFree(
+            ArrayBuilder<string> elementNames, ArrayBuilder<string> inferredElementNames)
+        {
+            if (elementNames == null && inferredElementNames == null)
+            {
+                return (default(ImmutableArray<string>), default(ImmutableArray<bool>));
+            }
+            if (elementNames == null && inferredElementNames != null)
+            {
+                var finalNames = inferredElementNames.ToImmutableAndFree();
+                return (finalNames, finalNames.SelectAsArray(n => n != null));
+            }
+            if (elementNames != null && inferredElementNames == null)
+            {
+                return (elementNames.ToImmutableAndFree(), default(ImmutableArray<bool>));
+            }
+
+            Debug.Assert(elementNames.Count == inferredElementNames.Count);
+            var builder = ArrayBuilder<bool>.GetInstance(elementNames.Count);
+            for (int i = 0; i < elementNames.Count; i++)
+            {
+                string inferredName = inferredElementNames[i];
+                if (elementNames[i] == null && inferredName != null)
+                {
+                    elementNames[i] = inferredName;
+                    builder.Add(true);
+                }
+                else
+                {
+                    builder.Add(false);
+                }
+            }
+
+            return (elementNames.ToImmutableAndFree(), builder.ToImmutableAndFree());
+        }
+
+        private static void RemoveDuplicateInferredTupleNames(ArrayBuilder<string> inferredElementNames, PooledHashSet<string> uniqueFieldNames)
+        {
+            if (inferredElementNames == null)
+            {
+                return;
+            }
+
+            for (int i = 0; i < inferredElementNames.Count; i++)
+            {
+                string name = inferredElementNames[i];
+                if (!uniqueFieldNames.Add(name))
+                {
+                    // remove all instances of this inferred name
+                    for (int j = 0; j < inferredElementNames.Count; j++)
```

Comment:
Can the inner loop be replaced with `inferredElementNames[i] = null;`? That would avoid potentially quadratic behavior and would avoid duplicating the `StringComparison` type here and in the `HashSet<string>`.